### PR TITLE
feat: add the multipeer connectivity transport

### PR DESCRIPTION
**Pull request by D4ryl00**
Add the multipeer connectivity transport of berty v1 to the master branch of berty
* [x] add the transport + driver in an internal package

### CODE REVIEW

File: go/internal/multipeer-connectivity-transport/driver/logger.gen.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Code generated by berty.tech/core/.scripts/generate-logger.sh
```
Comment (moul) [inline go/internal/multipeer-connectivity-transport/driver/logger.gen.go:1]: we're not using this method anymore, but I can help you get rid of this

File: go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -0,0 +1,297 @@
+// !$*UTF8*$!
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.xcodeproj/project.pbxproj:1]: Why did you add this file?

File: go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:1]: Why did you add a XCode workspace in the Go package?

Comment (aeddi) [reply go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:1]: I comment only this file but this comment is for the entire folder

File: go/internal/multipeer-connectivity-transport/multiaddr/logger.go [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package multiaddr
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/multiaddr/logger.go:1]: Do you need to have a logger in this package?

File: go/internal/multipeer-connectivity-transport/transport.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	host "github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	tptu "github.com/libp2p/go-libp2p-transport-upgrader"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/pkg/errors"
+)
+
+const DefaultBind = "/mc/Qmeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"
+
+// Transport is a BLE tpt.transport.
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/transport.go:20]:
```suggestion
// Transport is a MPConnectivity tpt.transport.
```

File: go/internal/multipeer-connectivity-transport/transport.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+var _ tpt.Transport = &Transport{}
+
+// Transport represents any device by which you can connect to and accept
+// connections from other peers.
+type Transport struct {
+	host     host.Host
+	upgrader *tptu.Upgrader
+}
+
+// NewTransport creates a BLE transport object that tracks dialers and listener.
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/transport.go:30]:
```suggestion
// NewTransport creates a MPConnectivity transport object that tracks dialers and listener.
```

File: go/internal/multipeer-connectivity-transport/transport.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+// It also starts the discovery service.
+func NewTransport(h host.Host, u *tptu.Upgrader) (*Transport, error) {
+	return &Transport{
+		host:     h,
+		upgrader: u,
+	}, nil
+}
+
+// Dial dials the peer at the remote address.
+// With BLE you can only dial a device that is already connected with the native driver.
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/transport.go:40]:
```suggestion
// With MPConnectivity you can only dial a device that is already connected with the native driver.
```

File: go/internal/multipeer-connectivity-transport/transport.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+func (t *Transport) Dial(ctx context.Context, remoteMa ma.Multiaddr,
+	remotePID peer.ID) (tpt.CapableConn, error) {
+	// BLE transport needs to have a running listener in order to dial other peer
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/transport.go:43]:
```suggestion
	// MPConnectivity transport needs to have a running listener in order to dial other peer
```

File: go/internal/multipeer-connectivity-transport/transport.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+	// because native driver is initialized during listener creation.
+	if gListener == nil {
+		return nil, errors.New("transport dialing peer failed: no active listener")
+	}
+
+	// remoteAddr is supposed to be equal to remotePID since with BLE transport:
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/transport.go:49]:
```suggestion
	// remoteAddr is supposed to be equal to remotePID since with MPConnectivity transport:
```

File: go/internal/multipeer-connectivity-transport/transport.go [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+	// multiaddr = /mc/<peerID>
+	remoteAddr, err := remoteMa.ValueForProtocol(mcma.P_MC)
+	if err != nil || remoteAddr != remotePID.Pretty() {
+		return nil, errors.Wrap(err, "transport dialing peer failed: wrong multiaddr")
+	}
+
+	// Check if native driver is already connected to peer's device.
+	// With BLE you can't really dial, only auto-connect with peer nearby.
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/transport.go:57]:
```suggestion
	// With MPConnectivity you can't really dial, only auto-connect with peer nearby.
```

File: go/internal/multipeer-connectivity-transport/connmgr.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+)
+
+// Connmgr keeps tracks of opened conn so the native driver can read from them
+// and close them.
+var connMap sync.Map
+
+// newConn returns an inbound or outbound tpt.CapableConn upgraded from a Conn.
+func newConn(ctx context.Context, t *Transport, remoteMa ma.Multiaddr,
+	remotePID peer.ID, inbound bool) (tpt.CapableConn, error) {
+	// Creates a BLE manet.Conn
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/connmgr.go:23]:
```suggestion
	// Creates a MPConnectivity manet.Conn
```

File: go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver/mc-driver.m [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    });
+    return gMCManager;
+}
+
+int StartMCDriver(char *localPID) {
+    NSLog(@"StartMCDriver called with: %s", localPID);
+    if (!driverStarted) {
+        NSString *cPID = [[NSString alloc] initWithUTF8String:localPID];
+        if (!getMCManager(cPID)) {
+            NSLog(@"StartBleDriver failed");
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver/mc-driver.m:35]:
```suggestion
            NSLog(@"StartMCDriver failed");
```

File: go/internal/multipeer-connectivity-transport/listener.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	tpt "github.com/libp2p/go-libp2p-core/transport"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// Global listener is used by discovery (to send incoming conn request to Accept())
+// and transport (to ensure that only one listener is running at a time).
+var gListener *Listener
+
+// Listener is a BLE tpt.Listener.
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/listener.go:20]:
```suggestion
// Listener is a MPConnectivity tpt.Listener.
```

File: go/internal/multipeer-connectivity-transport/listener.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+// Accept waits for and returns the next connection to the listener.
+// Returns a Multiaddr friendly Conn.
+func (l *Listener) Accept() (tpt.CapableConn, error) {
+	for {
+		select {
+		case req := <-l.inboundConnReq:
+			conn, err := newConn(l.ctx, l.transport, req.remoteMa, req.remotePID, true)
+			// If the BLE handshake failed for some reason, Accept won't return an error
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/listener.go:72]:
```suggestion
			// If the MPConnectivity handshake failed for some reason, Accept won't return an error
```

File: go/internal/multipeer-connectivity-transport/multiaddr/init.go [hunk 15]
```diff
@@ -0,0 +1,13 @@
+package multiaddr
+
+import (
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// Add BLE to the list of libp2p's multiaddr protocols
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/multiaddr/init.go:7]:
```suggestion
// Add MPConnectivity to the list of libp2p's multiaddr protocols
```

File: go/internal/multipeer-connectivity-transport/multiaddr/multiaddr.go [hunk 16]
```diff
@@ -0,0 +1,19 @@
+package multiaddr
+
+import (
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// BLE multiaddr protocol definition
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/multiaddr/multiaddr.go:7]:
```suggestion
// MPConnectivity multiaddr protocol definition
```

File: go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist (hunk 3 above)
Comment (D4ryl00) [reply go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:1]: > Why did you add a XCode workspace in the Go package?

Because that was that we did with the ble android package. I thought I should keep it because it's more easy to modify this package with the xcode workspace for everyone.

Comment (aeddi) [reply go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:1]: In v1, the gradle files were included in the Android driver since they were necessary to compile it.

On the other hand, there was no XCode file in the BLE Android driver since XCode was not used and cgo compiles it directly. It is the same for the MPConnectivity driver.

File: go/internal/multipeer-connectivity-transport/driver/logger.gen.go (hunk 1 above)
Comment (D4ryl00) [reply go/internal/multipeer-connectivity-transport/driver/logger.gen.go]: I've seen in other packages (e.g tinder) that logger is send from main.
For my package, I don't know how to do easily to give him the logger from main because it's a libp2p's Transporter constructor that constructs my transport.

**Comment by D4ryl00** [comment]
NewTransport won't failed if StartMCDriver failed.
Ready for review for merge.

File: go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.h [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+//  mc-driver.h
+//  mc-driver
+//
+//  Created by Rémi BARBERO on 30/03/2020.
+//  Copyright © 2020 Rémi BARBERO. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+
+int StartMCDriver(char *localPId);
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.h:11]: You should return void here if the returns is ignored by libp2p transport

File: go/internal/multipeer-connectivity-transport/driver/bridge.go [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+// Native -> Go functions
+func BindNativeToGoFunctions(hfp func(string) bool, rfp func(string, []byte)) {
+	native.GoHandleFoundPeer = hfp
+	native.GoReceiveFromPeer = rfp
+}
+
+// Go -> Native functions
+func StartMCDriver(localPID string) bool {
```
Comment (aeddi) [inline go/internal/multipeer-connectivity-transport/driver/bridge.go:16]: You shouldn't return a bool here if the return of this func is ignored by the transport

Comment (D4ryl00) [reply go/internal/multipeer-connectivity-transport/driver/bridge.go]: done!

File: go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.h (hunk 17 above)
Comment (D4ryl00) [reply go/internal/multipeer-connectivity-transport/driver/mc-driver/mc-driver.h]: done!
